rasa/engine/training/graph_trainer.py

```diff
@@ -0,0 +1,161 @@
+import copy
+import logging
+from pathlib import Path
+from typing import Any, Dict, Text, Type, Union
+
+from rasa.engine.caching import TrainingCache
+from rasa.engine.graph import ExecutionContext, GraphSchema
+from rasa.engine.runner.interface import GraphRunner
+from rasa.engine.storage.storage import ModelStorage
+from rasa.engine.training.components import (
+    CachedComponent,
+    FingerprintComponent,
+    FingerprintStatus,
+)
+from rasa.engine.training.hooks import TrainingHook
+from rasa.shared.core.domain import Domain
+
+
+logger = logging.getLogger(__name__)
+
+
+class GraphTrainer:
+    """Trains a model using a graph schema."""
+
+    def __init__(
+        self,
+        model_storage: ModelStorage,
+        cache: TrainingCache,
+        graph_runner_class: Type[GraphRunner],
+    ) -> None:
+        """Initializes a `GraphTrainer`.
+
+        Args:
+            model_storage: Storage which graph components can use to persist and load.
+                Also used for packaging the trained model.
+            cache: Cache used to store fingerprints and outputs.
+            graph_runner_class: The class to instantiate the runner from.
+        """
+        self._model_storage = model_storage
+        self._cache = cache
+        self._graph_runner_class = graph_runner_class
+
+    def train(
+        self,
+        train_schema: GraphSchema,
+        predict_schema: GraphSchema,
+        domain_path: Path,
+        output_filename: Path,
+    ) -> GraphRunner:
+        """Trains and packages a model then returns the prediction graph runner.
+
+        Args:
+            train_schema: The train graph schema.
+            predict_schema: The predict graph schema.
+            domain_path: The path to the domain file.
+            output_filename: The location to save the packaged model.
+
+        Returns:
+            A graph runner loaded with the predict schema.
+
+        """
+        logger.debug("Starting training.")
+
+        pruned_training_schema = self._fingerprint_and_prune(train_schema)
+
+        hooks = [TrainingHook(cache=self._cache, model_storage=self._model_storage)]
+
+        graph_runner = self._graph_runner_class.create(
+            graph_schema=pruned_training_schema,
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(graph_schema=pruned_training_schema),
+            hooks=hooks,
+        )
+
+        logger.debug("Train run.")
+
+        graph_runner.run()
+
+        domain = Domain.from_path(domain_path)
```

Comment:
this seems slightly weird - should we pass in a `Domain` object instead of the `Path`?

Reply:
You reckon that's nicer?

Reply:
keeps some responsibilities of `graph_trainer` 🤷🏻‍♂️ both options are fine to me